cli/cmd/uninstall/uninstall.go

```diff
@@ -361,10 +427,106 @@ func (c *Command) uninstallHelmRelease(releaseName, namespace, releaseType strin
 		c.UI.Output("Uninstall result: %s", res.Info, terminal.WithInfoStyle())
 		return nil
 	}
+
 	c.UI.Output(fmt.Sprintf("Successfully uninstalled %s Helm release.", releaseType), terminal.WithSuccessStyle())
 	return nil
 }
 
+// fetchCustomResources gets a list of all custom resources deployed in the
+// cluster that are managed by Consul.
+func (c *Command) fetchCustomResources() ([]unstructured.Unstructured, error) {
+	crs := make([]unstructured.Unstructured, 0)
+
+	crds, err := c.apiext.ApiextensionsV1().CustomResourceDefinitions().List(c.Ctx, metav1.ListOptions{
+		LabelSelector: "app=consul",
+	})
+	if err != nil {
+		return crs, err
+	}
+
+	for _, crd := range crds.Items {
+		for _, version := range crd.Spec.Versions {
+			target := schema.GroupVersionResource{
+				Group:    crd.Spec.Group,
+				Version:  version.Name,
+				Resource: crd.Spec.Names.Plural,
+			}
+
+			crList, err := c.dynamic.Resource(target).List(c.Ctx, metav1.ListOptions{})
+			if common.IgnoreNotFoundError(err) != nil {
+				return nil, err
+			}
+			if crList != nil {
+				crs = append(crs, crList.Items...)
+			}
+		}
+	}
+
+	return crs, nil
+}
+
+// deleteCustomResources takes a list of unstructured custom resources and
+// sends a request to each one to be deleted.
+func (c *Command) deleteCustomResources(crs []unstructured.Unstructured) error {
+	for _, cr := range crs {
+		apiVersion := strings.Split(cr.GetAPIVersion(), "/")
+		group, version := apiVersion[0], apiVersion[1]
+		if group == "" || version == "" {
+			return fmt.Errorf("malformed api version: %s", apiVersion)
+		}
+
+		target := schema.GroupVersionResource{
+			Group:    group,
+			Version:  version,
+			Resource: strings.ToLower(cr.GetKind()),
+		}
+
+		err := c.dynamic.
+			Resource(target).
+			Namespace(cr.GetNamespace()).
+			Delete(c.Ctx, cr.GetName(), metav1.DeleteOptions{})
+		if common.IgnoreNotFoundError(err) != nil {
```

Comment:
We ignore the not found error here to handle the case where a previous deletion request has deleted the CR, but the CR was still present when the last `fetchCustomResources` function call was made. As an example:

```
Initial CRs [A, B, C]
CR Deletion Attempt 1 {
    Fetch 1 returns [A, B, C]
    Deletion request sent to [A, B, C]
    A successfully deleted.
    B successfully deleted.
    Fetch 2 returns [C]
}
CR Deletion Attempt 2 {
    Fetch 1 returns [C]
    C successfully deleted (from deletion request in attempt 1).
    Deletion request sent to [C].
    C cannot be found (this is ignored).
    Fetch 2 returns []
}
Deletion attempts complete as Fetch 2 returned [].
```